code/datums/components/nanites.dm

```diff
@@ -0,0 +1,163 @@
+	var/list/programs_to_remove = programs.Copy()
+	var/list/programs_to_add = source.programs.Copy()
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		for(var/Y in programs_to_add)
+			var/datum/nanite_program/SNP = Y
+			if(NP.type == SNP.type)
+				programs_to_remove -= NP
+				programs_to_add -= SNP
+				SNP.copy_programming(NP, copy_activation)
+				break
+	if(full_overwrite)
+		for(var/X in programs_to_remove)
+			qdel(X)
+	for(var/X in programs_to_add)
+		var/datum/nanite_program/SNP = X
+		add_program(SNP.copy())
+
+/datum/component/nanites/proc/cloud_sync()
+	if(!cloud_id)
+		return
+	var/datum/nanite_cloud_backup/backup = SSnanites.get_cloud_backup(cloud_id)
+	if(backup)
+		var/datum/component/nanites/cloud_copy = backup.nanites
+		if(cloud_copy)
+			sync(cloud_copy)
+
+/datum/component/nanites/proc/add_program(datum/nanite_program/new_program, datum/nanite_program/source_program)
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		if(NP.unique && NP.type == new_program.type)
+			qdel(NP)
+	if(programs.len >= max_programs)
+		return
+	if(source_program)
+		source_program.copy_programming(new_program)
+	programs += new_program
+	new_program.on_add(src)
+
+/datum/component/nanites/proc/consume_nanites(amount, force = FALSE)
+	if(!host_mob) //dormant
+		return FALSE
+	if(!force && safety_threshold && (nanite_volume - amount < safety_threshold))
+		return FALSE
+	adjust_nanites(-amount)
+	return (nanite_volume > 0)
+
+/datum/component/nanites/proc/adjust_nanites(amount)
+	nanite_volume = CLAMP(nanite_volume + amount, 0, max_nanites)
+
+/datum/component/nanites/proc/on_emp(severity)
+	nanite_volume *= (rand(0.60, 0.90))		//Lose 10-40% of nanites
+	adjust_nanites(-(rand(5, 50)))		//Lose 5-50 flat nanite volume
+	if(prob(40/severity))
+		cloud_id = 0
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_emp(severity)
+
+/datum/component/nanites/proc/on_shock(shock_damage)
+	nanite_volume *= (rand(0.45, 0.80))		//Lose 20-55% of nanites
+	adjust_nanites(-(rand(5, 50)))			//Lose 5-50 flat nanite volume
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_shock(shock_damage)
+
+/datum/component/nanites/proc/on_minor_shock()
+	adjust_nanites(-(rand(5, 15)))			//Lose 5-15 flat nanite volume
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_minor_shock()
+
+/datum/component/nanites/proc/on_death(gibbed)
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_death(gibbed)
+		
+/datum/component/nanites/proc/on_hear(message, atom/movable/speaker, message_language, raw_message, radio_freq, list/spans, message_mode)
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_hear(message, speaker, message_language, raw_message, radio_freq, spans, message_mode)
+
+/datum/component/nanites/proc/receive_signal(code, source = "an unidentified source")
+	if(!host_mob) //dormant
```

Comment:
Unregister the signal if you're not on a mob

Reply:
same as above i never register the signal if i'm not on a mob, this is more of a paranoid sanity check

Reply:
let it runtime